theano/tensor/nnet/ctc_wrapper.c

```diff
@@ -0,0 +1,242 @@
+                      ctc_msg );
+        *status = 1;
+    }
+    *status = 0;
+}
+
+void create_contiguous_input_lengths( PyArrayObject * input_lengths_arr,
+    int ** input_lengths )
+{
+    int num_elements = PyArray_DIMS( input_lengths_arr )[0];
+
+    *input_lengths = (int *) malloc( num_elements * sizeof(int) );
+
+    if ( NULL == (*input_lengths) )
+        return;
+
+    for( int elem_idx = 0; elem_idx < num_elements; ++elem_idx )
+    {
+        (*input_lengths)[elem_idx] = *( (int *) PyArray_GETPTR1( input_lengths_arr, elem_idx ) );
+    }
+}
+
+void create_flat_labels( PyArrayObject * label_matrix, int ** flat_labels, 
+    int ** label_lengths )
+{
+    int rows = PyArray_DIMS( label_matrix )[0];
+    int cols = PyArray_DIMS( label_matrix )[1];
+
+    *flat_labels = (int *) malloc( rows * cols * sizeof(int) );
+    if ( NULL == (*flat_labels) )
+        return;
+
+    *label_lengths = (int *) malloc( rows * sizeof(int) );
+    if ( NULL == (*label_lengths) )
+    {
+        free( *flat_labels );
+        *flat_labels = NULL;
+        return;
+    }
+
+    int label_index = 0;
+    for( int row_idx = 0; row_idx < rows; ++row_idx )
+    {
+        int label_length = 0;
+        for( int col_idx = 0; col_idx < cols; ++col_idx )
+        {
+            int label = *( (int *) PyArray_GETPTR2( label_matrix, row_idx, col_idx ) );
+            if ( label >= 0 )  // negative values are assumed to be padding
+            {
+                (*flat_labels)[ label_index++ ] = label;
+                ++label_length;
+            }
+        }
+        (*label_lengths)[ row_idx ] = label_length;
+    }
+}
+
+#section support_code_apply
+
+int APPLY_SPECIFIC(ctc_cost_cpu)(PyArrayObject *  in_activations,
+                                 PyArrayObject *  in_labels,
+                                 PyArrayObject *  in_input_lengths,
+                                 PyArrayObject ** out_costs,
+                                 PyArrayObject ** out_gradients)
+{
+    npy_float32 * activations = NULL;
+    PyArrayObject * activations_copy = NULL;
+
+    if ( PyArray_IS_C_CONTIGUOUS( in_activations ) )
+    {
+        activations = (npy_float32 *) PyArray_DATA( in_activations );
+    }
+    else
+    {
+        activations_copy = PyArray_GETCONTIGUOUS( in_activations );
+        if ( NULL != activations_copy )
+        {
+            activations = (npy_float32 *) PyArray_DATA( activations_copy );
+        }
+        else
+        {
+            PyErr_Format( PyExc_MemoryError,
+                "Could not create a contiguous copy of activations array." );
+            return 1;            
+        }
+    }
+
+    int * input_lengths = NULL,
+        * flat_labels = NULL,
+        * label_lengths = NULL;
+
+    create_contiguous_input_lengths( in_input_lengths, &input_lengths );
+
+    if ( NULL == input_lengths )
+    {
+        PyErr_Format( PyExc_MemoryError,
+            "Could not allocate storage for input lengths" );
+        return 1;        
```

Comment:
`activations_copy` may have been allocated at this stage, too !

To help manage memory allocation/deallocation and avoid memory leaks, what I can suggest is to create a struct that will contain all pointers that you may want to allocate and deallocate. For example, a struct called `CPUCtctCostContext` with an init function (`CPUCtctCostContext_init(CPUCtctCostContext*)` that set any pointer member to NULL), and a clear function (`CPUCtctCostContext_clear(CPUCtctCostContext*)`, which free any pointer member that is not null). You can put this struct and its functions in `#section support_code` section.

Then, in `APPLY_SPECIFIC(ctc_cost_cpu)`, you create the struct on stack (`CPUCtctCostContext context;`), you initialize it (`CPUCtctCostContext_init(&context)`), you allocate members when you need (e.g. `context.activations_copy = ...;`) and then, everywhere an error occurs (like here), you just call `CPUCtctCostContext_clear(&context)` just before returning.